From 698463317c81ab1561a22a95540d48457838dcbc Mon Sep 17 00:00:00 2001
From: yongari <yongari@FreeBSD.org>
Date: Mon, 26 Nov 2012 04:40:26 +0000
Subject: [PATCH 12/12] MFC r242426:   TCP/UDP checksum offloading feature for
 IP fragmented datagram was   removed in r99417. 
 bge(4) controllers can do TCP checksum offload   for
 IP fragmented datagrams but unlike ti(4), it lacks
 UDP checksum   offloading for IP fragmented
 datagrams. The problem was bge(4)   blindly requested
 TCP/UDP checksum for IP fragmented datagrams such  
 that it resulted in corrupted UDP datagrams before
 r99417.   Remove remaining code for TCP checksum
 offloading for IP fragmented   datagrams which should
 have been removed in r99417.

---
 sys/dev/bge/if_bge.c |   27 ---------------------------
 1 file changed, 27 deletions(-)

diff --git a/sys/dev/bge/if_bge.c b/sys/dev/bge/if_bge.c
index e106cde..8f365e2 100644
--- a/sys/dev/bge/if_bge.c
+++ b/sys/dev/bge/if_bge.c
@@ -5128,10 +5128,6 @@ bge_encap(struct bge_softc *sc, struct mbuf **m_head, uint32_t *txidx)
 				return (error);
 			}
 		}
-		if (m->m_flags & M_LASTFRAG)
-			csum_flags |= BGE_TXBDFLAG_IP_FRAG_END;
-		else if (m->m_flags & M_FRAG)
-			csum_flags |= BGE_TXBDFLAG_IP_FRAG;
 	}
 
 	if ((m->m_pkthdr.csum_flags & CSUM_TSO) == 0) {
@@ -5253,29 +5249,6 @@ bge_start_locked(struct ifnet *ifp)
 			break;
 
 		/*
-		 * XXX
-		 * The code inside the if() block is never reached since we
-		 * must mark CSUM_IP_FRAGS in our if_hwassist to start getting
-		 * requests to checksum TCP/UDP in a fragmented packet.
-		 *
-		 * XXX
-		 * safety overkill.  If this is a fragmented packet chain
-		 * with delayed TCP/UDP checksums, then only encapsulate
-		 * it if we have enough descriptors to handle the entire
-		 * chain at once.
-		 * (paranoia -- may not actually be needed)
-		 */
-		if (m_head->m_flags & M_FIRSTFRAG &&
-		    m_head->m_pkthdr.csum_flags & (CSUM_DELAY_DATA)) {
-			if ((BGE_TX_RING_CNT - sc->bge_txcnt) <
-			    m_head->m_pkthdr.csum_data + 16) {
-				IFQ_DRV_PREPEND(&ifp->if_snd, m_head);
-				ifp->if_drv_flags |= IFF_DRV_OACTIVE;
-				break;
-			}
-		}
-
-		/*
 		 * Pack the data into the transmit ring. If we
 		 * don't have room, set the OACTIVE flag and wait
 		 * for the NIC to drain the ring.
-- 
1.7.9.5

